# Final Project Part B (100 points - Due on Sunday May 15th by 11:00 pm ET)
The output of this assignment for submission should be in PDF format AND .py or .ipynb. The name of the file should be as follows: Lastname_Firstname_Final_PartB.pdf (example: Thomas_Sunela_Final_PartB.pdf) AND Lastname_Firstname_Final_PartB.ipynb (example: Thomas_Sunela_Final_PartB.ipynb. In short, you are submitting the python notebook as well as the pdf of that notebook. Make sure your PDF file displays the visuals (if any) prior to submitting.

Incorrect file name, file type and PDF not displaying visuals (if any required) will cost you points!
Instructions for converting a Jupyter Python notebook to PDF: Go to the menu and choose, File --> Download As --> html

Open that html file and print it to PDF.

Use the YELP dataset provided in the shared google drive at: https://drive.google.com/drive/folders/1AAGizX5MzcnM8apuukUYg-37tUH4dz2P?usp=sharing

## Yelp Data

For the final project, you will analyze data that has been made publicly available for educational and academic purposes. The data contains 3 main tables we will analyze:

  + `yelp_academic_dataset_business.csv` - contains information about  a subset of businesses having at least one Yelp review. Every distinct business has a unique `business_id`
  
  + `yelp_academic_dataset_review.csv` - contains a subset of Yelp reviews with a `business_id` indicating what business the review is for and a `user_id` to identify the user leaving the review
  
  + `yelp_academic_dataset_user.csv` - contains a subset of Yelp users identified by `user_id`
   
## What you need to do

Your goal in this notebook is to work through the problems below to perform some brief analyses of the `yelp_academic_dataset_business.csv` data. 

As always, it can be useful to document your thoughts, strategy, and assumptions as you go. If you are stuck on a particular problem, write down what you are trying to do and why you think you may be getting stuck.

## Setup - Run the following code to get started

In [1]:
# Options to display more rows and columns
# DO NOT REMOVE

import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 100)

In [2]:
# Load the libraries you need...
import numpy as np
%matplotlib inline
import seaborn as sns
import matplotlib.pyplot as plt


In [12]:
# Load the data

business_data = pd.read_csv(r'yelp_academic_dataset_business.csv')
business_data.head()

address  attributes attributes.AcceptsInsurance attributes.AgesAllowed  \
0    1314 44 Avenue NE         NaN                         NaN                    NaN   
1                  NaN         NaN                         NaN                    NaN   
2  1335 rue Beaubien E         NaN                         NaN                    NaN   
3      211 W Monroe St         NaN                         NaN                    NaN   
4  2005 Alyth Place SE         NaN                         NaN                    NaN   

  attributes.Alcohol                                attributes.Ambience attributes.BYOB  \
0                NaN                                                NaN             NaN   
1               none                                                NaN             NaN   
2      beer_and_wine  {'romantic': False, 'intimate': False, 'classy...             NaN   
3                NaN                                                NaN             NaN   
4                NaN                                                NaN             NaN   

  attributes.BYOBCorkage attributes.BestNights attributes.BikeParking  \
0                    NaN                   NaN                  False   
1                    NaN                   NaN                  False   
2                    NaN                   NaN                   True   
3                    NaN                   NaN                    NaN   
4                    NaN                   NaN                    NaN   

  attributes.BusinessAcceptsBitcoin attributes.BusinessAcceptsCreditCards  \
0                               NaN                                  True   
1                               NaN                                  True   
2                               NaN                                 False   
3                               NaN                                   NaN   
4                               NaN                                  True   

                          attributes.BusinessParking attributes.ByAppointmentOnly  \
0  {'garage': False, 'street': True, 'validated':...                          NaN   
1  {'garage': False, 'street': True, 'validated':...                          NaN   
2  {'garage': False, 'street': False, 'validated'...                          NaN   
3                                                NaN                          NaN   
4                                                NaN                          NaN   

  attributes.Caters attributes.CoatCheck attributes.Corkage attributes.DietaryRestrictions  \
0               NaN                  NaN                NaN                            NaN   
1              True                  NaN                NaN                            NaN   
2             False                  NaN                NaN                            NaN   
3               NaN                  NaN                NaN                            NaN   
4               NaN                  NaN                NaN                            NaN   

  attributes.DogsAllowed attributes.DriveThru attributes.GoodForDancing attributes.GoodForKids  \
0                    NaN                  NaN                       NaN                   True   
1                   True                False                       NaN                   True   
2                    NaN                  NaN                       NaN                   True   
3                    NaN                  NaN                       NaN                    NaN   
4                    NaN                  NaN                       NaN                    NaN   

                              attributes.GoodForMeal attributes.HairSpecializesIn  \
0                                                NaN                          NaN   
1  {'dessert': False, 'latenight': False, 'lunch'...                          NaN   
2  {'dessert': False, 'latenight': False, 'lunch'...                          NaN   
3                         

## Task Set #1 - Explore and Clean

### Problem 1: (5 points)

<span  style="color:purple; font-size:16px">The `yelp_academic_dataset_business.csv` data has been loaded for you as a Pandas DataFrame called `business_data`. Determine the  number of rows and columns of the data. **Is the data in a tidy format? Briefly explain your reasoning about why you think the data is/is not tidy.**</span>

In [13]:
# Your code here
print('number of rows =',business_data.shape[0])
print('number of columns =',business_data.shape[1])

number of rows = 188593
number of columns = 61


#### Written Answer:

number of rows = 188593


number of columns = 61



### Problem 2: (5 points)

<span  style="color:purple; font-size:16px">Determine the number of columns of each data type in `business_data`: object, float64, int64. </span>

In [14]:
# Your code here
business_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188593 entries, 0 to 188592
Data columns (total 61 columns):
 #   Column                                 Non-Null Count   Dtype  
---  ------                                 --------------   -----  
 0   address                                180970 non-null  object 
 1   attributes                             0 non-null       float64
 2   attributes.AcceptsInsurance            11671 non-null   object 
 3   attributes.AgesAllowed                 397 non-null     object 
 4   attributes.Alcohol                     47892 non-null   object 
 5   attributes.Ambience                    47577 non-null   object 
 6   attributes.BYOB                        911 non-null     object 
 7   attributes.BYOBCorkage                 1409 non-null    object 
 8   attributes.BestNights                  6844 non-null    object 
 9   attributes.BikeParking                 84891 non-null   object 
 10  attributes.BusinessAcceptsBitcoin      12674 non-null   

### Problem 3: (5 points)

<span  style="color:purple; font-size:16px">Create a Pandas Series containing the proportion of missing values in each column of `business_data`.</span>

In [15]:
# Your code here
proportion_missing = (business_data.isnull().sum() * 100 / len(business_data)).round(2)
missing_value_series = pd.Series( proportion_missing,business_data.columns)

missing_value_series.head(100)

address                                    4.04
attributes                               100.00
attributes.AcceptsInsurance               93.81
attributes.AgesAllowed                    99.79
attributes.Alcohol                        74.61
attributes.Ambience                       74.77
attributes.BYOB                           99.52
attributes.BYOBCorkage                    99.25
attributes.BestNights                     96.37
attributes.BikeParking                    54.99
attributes.BusinessAcceptsBitcoin         93.28
attributes.BusinessAcceptsCreditCards     25.56
attributes.BusinessParking                45.16
attributes.ByAppointmentOnly              75.91
attributes.Caters                         78.77
attributes.CoatCheck                      95.48
attributes.Corkage                        99.65
attributes.DietaryRestrictions            99.93
attributes.DogsAllowed                    92.75
attributes.DriveThru                      96.42
attributes.GoodForDancing               

In [ ]:
##option two

# business_data.isnull().sum(axis='columns').value_counts(normalize=True)

<span  style="color:purple; font-size:16px">How many variables have no missing values?</span>


There are 6 variables that has no missing values.The variables are:

business_id  
is_open                                                                     
name                                       
review_count                               
stars                                     
state                                      

In [16]:
#your code here
# business_data.isnull().all()
missing_value_columns= business_data[business_data.columns[~business_data.isna().any()]]

missing_value_columns.head()

business_id  is_open                  name  review_count  stars state
0  Apn5Q_b6Nz61Tq4XzPdf9A        1  Minhas Micro Brewery            24    4.0    AB
1  AjEbIBw6ZFfln7ePHha9PA        0   CK'S BBQ & Catering             3    4.5    NV
2  O8S5hYJ1SMc8fA4QBtVujA        0         La Bastringue             5    4.0    QC
3  bFzdJJ3wp3PZssNEsyU23g        1       Geico Insurance             8    1.5    AZ
4  8USyCYqpScwiNEb58Bt6CA        1         Action Engine             4    2.0    AB

In [ ]:
# nomissingValues = missing_value_series[missing_value_series.columns[~df.isnull().all()]]

### Problem 4: (5 points)

<span  style="color:purple; font-size:16px">Use an appropriate visualization technique to analyze the missingness in `business_data`.</span>

In [ ]:
pip install missingno

In [ ]:
import seaborn as sns
import missingno as msno

In [ ]:
# Use a heatmap, pointplot, or other appropriate visualization here

msno.heatmap(business_data,figsize=(15,8))

In [ ]:
msno.bar(business_data)

In [ ]:

plt.figure(figsize=(15,6))
sns.displot(
    data=business_data.isna().melt(value_name="missing"),
    y="variable",
    hue="missing",
    multiple="fill",
    aspect=1.25
)
plt.show

### Problem 5: (5 points)

<span  style="color:purple; font-size:16px">Drop all columns that contain **only missing values** from `business_data`.</span>

In [33]:
# Your code here
business_data=business_data.dropna(axis='columns',how='all')

In [34]:
business_data.head(10)

address attributes_AcceptsInsurance attributes_AgesAllowed  \
0                1314 44 Avenue NE                         NaN                    NaN   
1                              NaN                         NaN                    NaN   
2              1335 rue Beaubien E                         NaN                    NaN   
3                  211 W Monroe St                         NaN                    NaN   
4              2005 Alyth Place SE                         NaN                    NaN   
5  20235 N Cave Creek Rd, Ste 1115                         NaN                    NaN   
6                   631 Bloor St W                         NaN                    NaN   
7      3417 Derry Road E, Unit 103                         NaN                    NaN   
8               1440 N. Dysart Ave                         NaN                    NaN   
9                  209 Oakland Ave                         NaN                    NaN   

  attributes_Alcohol                                attributes_Ambience attributes_BYOB  \
0                NaN                                                NaN             NaN   
1               none                                                NaN             NaN   
2      beer_and_wine  {'romantic': False, 'intimate': False, 'classy...             NaN   
3                NaN                                                NaN             NaN   
4                NaN                                                NaN             NaN   
5                NaN                                                NaN             NaN   
6                NaN                                                NaN             NaN   
7               none                                                NaN             NaN   
8               none  {'romantic': False, 'intimate': False, 'classy...             NaN   
9                NaN                                                NaN             NaN   

  attributes_BYOBCorkage attributes_BestNights attributes_BikeParking  \
0                    NaN                   NaN                  False   
1                    NaN                   NaN                  False   
2                    NaN                   NaN                   True   
3                    NaN                   NaN                    NaN   
4                    NaN                   NaN                    NaN   
5                    NaN                   NaN                   True   
6                    NaN                   NaN                    NaN   
7                    NaN                   NaN                    NaN   
8                    NaN                   NaN                   True   
9                    NaN                   NaN                   True   

  attributes_BusinessAcceptsBitcoin attributes_BusinessAcceptsCreditCards  \
0                               NaN                                  True   
1                               NaN                                  True   
2                               NaN                                 False   
3                               NaN                                   NaN   
4                               NaN                                  True   
5                               NaN                                  True   
6                               NaN                                   NaN   
7                               NaN                                  True   
8                               NaN                                  True   
9                               NaN                                  True   

                          attributes_BusinessParking attributes_ByAppointmentOnly  \
0  {'garage': False, 'street': True, 'validated':...                          NaN   
1  {'garage': False, 'street': True, 'validated':...                          NaN   
2  {'garage': False, 'street': False, 'validated'...                          NaN   
3                                                NaN 

### Problem 6: (5 points)

<span  style="color:purple; font-size:16px">The `business_data` is not in a form that is convenient for analysis. Let's begin by cleaning up the variable names. </span>

Tasks:

1. Make all column names lowercase.
1. Replace all periods `.` with an underscore _

In [35]:
# Your code here
business_data.columns= business_data.columns.str.replace('.','_',regex=False)
business_data.columns

Index(['address', 'attributes_AcceptsInsurance', 'attributes_AgesAllowed', 'attributes_Alcohol',
       'attributes_Ambience', 'attributes_BYOB', 'attributes_BYOBCorkage', 'attributes_BestNights',
       'attributes_BikeParking', 'attributes_BusinessAcceptsBitcoin',
       'attributes_BusinessAcceptsCreditCards', 'attributes_BusinessParking',
       'attributes_ByAppointmentOnly', 'attributes_Caters', 'attributes_CoatCheck',
       'attributes_Corkage', 'attributes_DietaryRestrictions', 'attributes_DogsAllowed',
       'attributes_DriveThru', 'attributes_GoodForDancing', 'attributes_GoodForKids',
       'attributes_GoodForMeal', 'attributes_HairSpecializesIn', 'attributes_HappyHour',
       'attributes_HasTV', 'attributes_Music', 'attributes_NoiseLevel', 'attributes_Open24Hours',
       'attributes_OutdoorSeating', 'attributes_RestaurantsAttire',
       'attributes_RestaurantsCounterService', 'attributes_RestaurantsDelivery',
       'attributes_RestaurantsGoodForGroups', 'attributes_Re

### Problem 7: (5 points)

<span  style="color:purple; font-size:16px">Let's now get rid of columns that we don't need right now. Perform the following transformation steps to the data.</span>

1. Drop all `attribute_*` columns except for the `attributes_happyhour` column.
1. Create a new column, called `open_on_weekends`, that is a 1 if the business is open on the weekend (has a non-null value for either `hours_saturday` or `hours_sunday`) and 0 otherwise.
1. Drop all of the `hours_*` columns after creating the `open_on_weekends` column.

In [36]:
# Drop all of the unwanted attributes columns
business_data_df = business_data.loc[:,~business_data.columns.str.startswith("attributes_")]
# adding the 'attributes_happyhour' back
business_data_df['attributes_HappyHour']=business_data['attributes_HappyHour']

/var/folders/_9/pz2ym61s237668kpj95dnxnw0000gn/T/ipykernel_95796/1818845138.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  business_data_df['attributes_HappyHour']=business_data['attributes_HappyHour']


In [37]:
business_data_df.head()

address             business_id                                         categories  \
0    1314 44 Avenue NE  Apn5Q_b6Nz61Tq4XzPdf9A  Tours, Breweries, Pizza, Restaurants, Food, Ho...   
1                  NaN  AjEbIBw6ZFfln7ePHha9PA  Chicken Wings, Burgers, Caterers, Street Vendo...   
2  1335 rue Beaubien E  O8S5hYJ1SMc8fA4QBtVujA  Breakfast & Brunch, Restaurants, French, Sandw...   
3      211 W Monroe St  bFzdJJ3wp3PZssNEsyU23g                      Insurance, Financial Services   
4  2005 Alyth Place SE  8USyCYqpScwiNEb58Bt6CA  Home & Garden, Nurseries & Gardening, Shopping...   

        city hours_Friday hours_Monday hours_Saturday hours_Sunday hours_Thursday hours_Tuesday  \
0    Calgary    11:0-21:0    8:30-17:0      11:0-21:0          NaN      11:0-21:0     11:0-21:0   
1  Henderson    17:0-23:0          NaN      17:0-23:0    17:0-23:0            NaN           NaN   
2   Montréal    10:0-22:0    10:0-22:0      10:0-22:0    10:0-22:0      10:0-22:0     10:0-22:0   
3    Phoenix          NaN          NaN            NaN          NaN            NaN           NaN   
4    Calgary     8:0-17:0     8:0-17:0            NaN          NaN       8:0-17:0      8:0-17:0   

  hours_Wednesday  is_open   latitude   longitude                  name  \
0       11:0-21:0        1  51.091813 -114.031675  Minhas Micro Brewery   
1             NaN        0  35.960734 -114.939821   CK'S BBQ & Catering   
2       10:0-22:0        0  45.540503  -73.599300         La Bastringue   
3             NaN        1  33.449999 -112.076979       Geico Insurance   
4        8:0-17:0        1  51.035591 -114.027366         Action Engine   

                neighborhood postal_code  review_count  stars state attributes_HappyHour  
0                        NaN     T2E 6L6            24    4.0    AB                  NaN  
1                        NaN       89002             3    4.5    NV                  NaN  
2  Rosemont-La Petite-Patrie     H2G 1K7             5    4.0    QC                  NaN  
3                        NaN       85003             8    1.5    AZ                  NaN  
4                        NaN     T2H 0N5             4    2.0    AB                  NaN

In [ ]:
# Create open_on_weekends

business_data_df = business_data_df.assign(open_on_weekends=np.where(business_data_df.hours_Saturday.isnull() & business_data_df.hours_Sunday.isnull()
                             , 0, 1))

In [45]:
# Remove hours.* columns
business_data_df = business_data_df.loc[:,~business_data_df.columns.str.startswith("hours_")]

In [46]:
business_data_df.head()

address             business_id                                         categories  \
0    1314 44 Avenue NE  Apn5Q_b6Nz61Tq4XzPdf9A  Tours, Breweries, Pizza, Restaurants, Food, Ho...   
1                  NaN  AjEbIBw6ZFfln7ePHha9PA  Chicken Wings, Burgers, Caterers, Street Vendo...   
2  1335 rue Beaubien E  O8S5hYJ1SMc8fA4QBtVujA  Breakfast & Brunch, Restaurants, French, Sandw...   
3      211 W Monroe St  bFzdJJ3wp3PZssNEsyU23g                      Insurance, Financial Services   
4  2005 Alyth Place SE  8USyCYqpScwiNEb58Bt6CA  Home & Garden, Nurseries & Gardening, Shopping...   

        city  is_open   latitude   longitude                  name               neighborhood  \
0    Calgary        1  51.091813 -114.031675  Minhas Micro Brewery                        NaN   
1  Henderson        0  35.960734 -114.939821   CK'S BBQ & Catering                        NaN   
2   Montréal        0  45.540503  -73.599300         La Bastringue  Rosemont-La Petite-Patrie   
3    Phoenix        1  33.449999 -112.076979       Geico Insurance                        NaN   
4    Calgary        1  51.035591 -114.027366         Action Engine                        NaN   

  postal_code  review_count  stars state attributes_HappyHour  open_on_weekends  
0     T2E 6L6            24    4.0    AB                  NaN                 1  
1       89002             3    4.5    NV                  NaN                 1  
2     H2G 1K7             5    4.0    QC                  NaN                 1  
3       85003             8    1.5    AZ                  NaN                 0  
4     T2H 0N5             4    2.0    AB                  NaN                 0

### Problem 8: (5 points)

<span  style="color:purple; font-size:16px">Create a Pandas Series containing the proportion of business in the data by state. (For now, a *state* refers to any level of the `state` variable, not only states in the U.S.A.)</span>

In [100]:
# Number of businesses by state
business_by_state=pd.Series(business_data_df[['state','categories']].value_counts(normalize=True)*100,).round(2)
business_by_state.head(100)

state  categories                                    
AZ     Restaurants, Mexican                              0.21
       Mexican, Restaurants                              0.20
       Beauty & Spas, Nail Salons                        0.17
ON     Coffee & Tea, Food                                0.17
       Restaurants, Chinese                              0.16
       Food, Coffee & Tea                                0.16
AZ     Nail Salons, Beauty & Spas                        0.16
ON     Chinese, Restaurants                              0.15
       Beauty & Spas, Hair Salons                        0.13
AZ     Hair Salons, Beauty & Spas                        0.13
       Auto Repair, Automotive                           0.12
       Beauty & Spas, Hair Salons                        0.12
ON     Hair Salons, Beauty & Spas                        0.12
AZ     Automotive, Auto Repair                           0.11
ON     Restaurants, Pizza                                0.11
AZ     Coffee & 

<span  style="color:purple; font-size:16px">How many unique states are represented in the data? Are all of them in the United States of America? Do all of the levels of `state` make sense?</span>

#### Written Answer:

*{Type your answer here}*



In [119]:
business_by_state.index.unique(level=0)

Index(['AZ', 'ON', 'NV', 'OH', 'PA', 'NC', 'QC', 'AB', 'WI', 'IL', 'SC', 'NYK', 'NI', 'IN', 'ST',
       'BY', 'CO', 'O', 'NLK', 'RP', 'XGM', 'OR', '01', 'NW', 'NY', 'PO', 'WHT', 'WAR', 'WA', 'VT',
       'VS', 'VA', 'V', 'TAM', 'SP', 'SG', 'RCC', 'AR', 'AL', 'AG', '11', '6', '45', '4', '10',
       'NE', 'CMA', 'CC', 'CA', 'C', 'GA', 'HE', 'G', 'FL', 'DE', 'CS', 'CRF', 'HU', 'HH', 'B',
       'BC', 'MT', 'MO', 'MN', 'MA', 'M', 'LU', 'XMS'],
      dtype='object', name='state')

### Problem 9: (5 points)

<span  style="color:purple; font-size:16px">These data are fairly large! Let's focus on the data from a few key states. The goal in this exercise is to subset `business_data` to only contain businesses in the 5 states having the largest total number of businesses in the state. (Again, *state* does not need to correspond to a state in the U.S.A. but simply refers to a level of the `state` column.)</span>

Tasks:

1. Extract the 5 states from `state_counts` with the largest number of businesses
2. Subset `business_data` to only contain observations from these top-5 states

In [ ]:
# Your code here


### Problem 10:  (5 points)

<span  style="color:purple; font-size:16px">Import the `territory_abbr.csv` data which contains a record for every U.S. state and territory. Use this data to **remove** businesses outside of the United States *by* performing an appropriate **merge** with `business_data`. Assign the remaining data to a DataFrame called `business_data_subset`.</span>

(Of course...you could probably figure out what state(s) need to be removed just by looking at them. The point is to demonstrate that you can use `merge()` and any necessary arguments to accomplish the task. Also...don't mess up your tidy data! Though not necessary for full credit, you may want to remove/modify columns that are not lowercase.)

In [ ]:
# Read in data

us_data = pd.read_csv('territory_abbr.csv')
us_data.head()

In [ ]:
# Your code here


<span  style="color:purple; font-size:16px">How many unique U.S. states are left in the data?</span>

In [ ]:
# Your code here


## Task Set #2 - Analysis

---

### Problem 11: (5 points)

<span  style="color:purple; font-size:16px">Do businesses with more reviews tend to have higher ratings? </span> 

<span  style="color:purple; font-size:16px">To help answer this question, use `business_data` to create *side-by-side notched boxplots* to visualize the relationship between the number of reviews left for the business (`review_count`, y-axis) and the rating of the business (`stars`, x-axis).</span> 
    
<span  style="color:purple; font-size:16px">**Note: use `ax.set_yscale('log')` to set the scale of the y-axis to be logarithmic due to extreme variation in the number of reviews left for the businesses. This means you must specify a figure and axes object using `fig, ax = plt.subplots(figsize=(10, 10))`**</span>
    
<span>Make sure to use appropriate axes labels and include an appropriate title. Make the figure 10 units tall and 10 units wide.</span>

In [ ]:
# Your code here



<span  style="color:purple; font-size:16px">Describe what you see in the plot and formulate a hypothesis about why the relationship between these variables looks the way it does.</span>

#### Written Answer:

*{Type your answer here}*



### Problem 12: (5 points)

<span  style="color:purple; font-size:16px">The `categories` column of `business_data_subset` contains a *comma separated* list of various keyword categories for each business. Use an appropriate Python data structure to iterate over every row of the data and determine all of the unique categories as well as the number of times each category appears, sorted from most common category to least common category.</span>

Your result should look like the following:

<pre>
Restaurants                  234
Shopping                     173
Home Services                132
Food                         120
Health & Medical              92
.
.
.

In [ ]:
# Your code here



<span  style="color:purple; font-size:16px">How many distinct categories are in `business_data_subset`?</span>

In [ ]:
# Number of distinct categories equal to the number of rows



### Problem 13: (5 points)

<span  style="color:purple; font-size:16px">For this question, we want to see if we can use the `categories` variable from the `business_data_subset` table to extract the records belonging to **food establishments** - any place that someone could go to get a freshly prepared meal.</span>

<span  style="color:purple; font-size:16px">Create a copy of the Pandas DataFrame called `food_establishments` which contains all business in `business_data_subset` that contain either *dining* or *restaurant* in their list of categories. (Make sure to ignore case.) How many food establishments are captured using this method?</span>

Hint: This can be accomplished in at least two ways. No reasonable method is bad! Create a new variable and use it to filter observations or directly subset the data. Your choice!

In [ ]:
# Your code here
look for the word restaurant or dinner

### Problem 14: (5 points)

<span  style="color:purple; font-size:16px">Using the `food_establishments` table, create a table with `stars` as the index, `state` as the columns, and the conditional `stars` distribution of food establishments for each `state` as values.</span>

<span  style="color:purple; font-size:16px">Include the marginal `stars` distribution in your table by using the appropriate `margins` argument. Create a valid conditional `stars` distribution for each state using the `normalize = 'columns'` argument.</span>

You can check that you have successfully create valid conditional `stars` distributions for each state by verifying each column sums to 1. 

Your final table should look similar to the table below with the appropriate values imputed for `val`.

| state  	|  AZ 	| NV  	| NC  	| OH  | All |
|---	|---	|---	|---	|---  |---  |
|   stars 	|
|   1 	|   val	|  val 	|  val 	| val   | val    |
|   1.5	|   val	|  val	|  val 	| val   | val    |
|   2	|   val	|  val 	|  val 	| val   | val    |
|   2.5	|   val	|  val	|  val 	| val   | val    |
|   3	|   val	|  val 	|  val 	| val   | val    |
|   3.5	|   val	|  val	|  val 	| val   | val    |
|   4	|   val	|  val 	|  val 	| val   | val    |
|   4.5	|   val	|  val 	|  val 	| val   | val    |
|   5	|   val	|  val 	|  val 	| val   | val    |

In [ ]:
# Your code here


### Problem 15: (5 points)

<span  style="color:purple; font-size:16px">Is there a difference in the rating distribution (`stars`) of food establishments by state?</span>

<span  style="color:purple; font-size:16px">Use the table you created in the previous problem to create an appropriate visualization that helps you to answer the above question.</span>

In [ ]:
# Your code here


### Problem 16: (5 points)

<span  style="color:purple; font-size:16px">Using the `food_establishments` data, calculate the *median*, *mean*, *standard deviation*, *minimum*, *maximum*, and *number of observations* for food establishments with/without a *Happy Hour* by using the `attributes_happyhour` variable.</span>

In [ ]:
# Your code here


<span  style="color:purple; font-size:16px">Describe any differences you find for food establishments having a *Happy Hour* and those that do not. Do food establishments with a happy hour seem to typically be higher or lower rated than food establishments without a happy hour? Do most food establishments offer a happy hour?</span>

#### Written Answer:

*{Type your answer here}*



### Problem 17: (5 points)

<span  style="color:purple; font-size:16px">Let's investigate differences in food establishments with/without a Happy Hour using an appropriate visualization.</span>
    
<span  style="color:purple; font-size:16px">First, create a new variable for the `food_establishments` table called `binned_review_count` that bins the `review_count` variable as: 
    
 + '[0-10]' = having between 0 and 10 reviews inclusive
 + '(10,100]' = having between 10 and 100 reviews
 + '(100,1,000]' = having between 100 and 1,000 reviews
 + '1,000 +` = more than one thousand reviews.
</span>

<span  style="color:purple; font-size:16px">Create a side-by-side violin plot with `binned_review_count` on the y-axis, `stars` on the x-axis, and allow the rating distributions to differ for establishments **with and without a Happy Hour** by using the `split = True` argument. As always, provide an appropriate title and axes labels.</span>

In [ ]:
# Your code here


Describe what you see in the plot.

#### Written Answer:

*{Type your answer here}*



### Problem 18: (5 points)

<span  style="color:purple; font-size:16px">Suppose you work at Yelp. You have been tasked with a new initiative to create a new award for the highest performing food establishments represented on the Yelp platform. The team at Yelp has determined that any food establishment that has both the highest `stars` rating **AND** the largest total number of reviews (`review_count`) in their city *deserves* the **Best Local Food Establishment Award**. You have been tasked with writing Python code that can determine which food establishments in a data set deserve this new prestigious award.</span>

<span  style="color:purple; font-size:16px">For this exercise, create a DataFrame containing all the food establishments in the `food_establishments` data that deserve **Best Local Food Establishment Award** for every city/state combination (to avoid issues with cities that have the same name). The final table that you are to create, called `best_food_establishments`, should provide the `name`, `city`, `state`, `review_count`, and `stars` ratings of all of the food establishments that deserve the award.</span>

Important Notes: There are some slight variations in some city names like: LasVegas, Las Vegas, las vegas. You do not need to worry about cleaning this up before performing your analysis, this is really a proof-of-concept for your team. Also, some cities will have no food establishments that deserve to receive this award as it is possible that the food establishment with the largest number of reviews **does not** have the highest rating in the city.

In [ ]:
# Your code here


### Problem 19: (10 points)

<span  style="color:purple; font-size:16px">Subsequent questions involve the `yelp_academic_dataset_review.csv` data. Let's investigate the relationship between the data for businesses remaining in `business_data` (we subset the data in a previous problem, you can use it as is) and the `yelp_academic_dataset_review.csv` data. One thing that would interesting to know is, does it appear that we have all the reviews for each business? For example: if we see a business in `business_data` has 14 reviews, does the review table contain all 14 of those reviews?</span>

<span  style="color:purple; font-size:16px">To determine whether this is the case:
    
  + Import the `yelp_academic_dataset_review.csv` data.
  + Subset the reviews data to only contain records for businesses in the `business_data` table.
  + Use `groupby()` to get the counts of reviews for every unique business (by `business_id`) in the reviews data.
  + Perform an appropriate join to the `business_data` table to create a DataFrame containing three columns: 
      1. `business_id`
      1. `number_reviews` - the number of reviews for that particular business you found in the review table, 
      1. `review_count` - the number of reviews listed in the `food_establishments` table
  + Create a new column `abs_diff` containing the absolute value of the difference in the number of reviews for each business in `business_data` and number of reviews from the reviews table.
  + Create a plot that helps us see the extent of any disparities between the number of reviews in the review data and the number of reviews stated in the `business_id` table.
  + Use the `describe()` method to numerically describe the extent of any disparities between the number of reviews in the `review_data` and the number of reviews stated in the `business_id` table in the `review_count` field.
</span>

In [ ]:
# Import the review data

review_data = pd.read_csv(r'yelp_academic_dataset_review.csv')
review_data.head()

In [ ]:
# Your code here


Describe what you see in the plot and the output from `describe()`. Does it appear that we have the all or most of the reviews for every business in the `business_data` table?

#### Written Answer:

*{Type your answer here}*



---

## Wrapping up

In this section of the Final Project, you will be focused on analyzing the Yelp users and the reviews they leave. Unlike the previous part, you will not be guided through each step and must decide how best to proceed.

Your analyses will involve the following 3 tables:

+ `yelp_academic_dataset_review.csv` - contains a subset of Yelp reviews with a `business_id` indicating what business the review is for and a `user_id` to identify the user leaving the review.

+ `yelp_academic_dataset_user.csv` - contains a subset of Yelp users identified by `user_id`

+ `yelp_academic_dataset_business.csv` - contains information about a subset of businesses having at least one Yelp review and every distinct business has a unique `business_id`. This is the table that we have been analyzing in Problems 1-19. 


In [ ]:
# Import the user data table

#user_data = pd.read_csv(r'data-sets/yelp_academic_dataset_user.csv')

### Problem 20: Time Based Analysis (20 points)

<span  style="color:purple; font-size:16px">The `review_data` table contains a column `date` representing the date the review was created.</span>

Chose **ONE** of the questions below and use the `review_data` to answer.

1. Are more reviews posted during certain days of the week? What about certain months of the year? What patterns can you find about when reviews are posted? Draw a suitable plot.

1. Do reviews made on certain days of the week or certain months of the year tend to be higher/lower rated? What relationship can you find between the `stars` ratings of a review and when the reviews was posted? Draw a suitable plot.

In [ ]:
# Your code here


<span  style="color:purple; font-size:16px">Provide a brief explanation of your conclusions.</span>

#### Written Answer:

*{Type your answer here}*



### Problem 21: Text Based Analysis (20 points)

<span  style="color:purple; font-size:16px">The `review_data` table contains a column `text` containing the full text of the review. Use these data to answer one of the questions below.</span>

Chose **ONE** of the questions below and use the `review_data` to answer.

1. Can you classify a review as having a *positive* or *negative* sentiment? If so, select five particular businesses (by `business_id`) and compare the business's average `stars` rating from the `business_data` to the proportion of reviews you have identified as having `positive` or `negative` sentiment for those particular businesses. Draw a suitable plot.

1. Do reviews with exclamation points seem to be either very highly rated or very low rated? Determine the `stars` distribution as a function of the number of exclamation points used in the review. Draw a suitable plot.


In [ ]:
# Your code here


<span  style="color:purple; font-size:16px">Provide a brief explanation of your conclusions.</span>

#### Written Answer:

*{Type your answer here}*



### Problem 22: Analysis of Users (20 points)

<span  style="color:purple; font-size:16px">Create a function that accepts at least a `user_id` as input. This function use the `yelp_academic_dataset_user.csv` data as well as the `yelp_academic_dataset_review.csv` or `yelp_academic_dataset_business.csv` tables as necessary.</span>

<span  style="color:purple; font-size:16px">The function should automatically process the data relevant to the `user_id` passed and create at least one plot annotated with relevant, interesting information about the user that best answers the following questions:</span>

**What types of businesses has the user reviewed? Summarize the users review ratings by business category.**

Tip: It is OK if the results of your function are not very interesting for users with only a small number of reviews but your function should produce something reasonable for any `user_id` that may be passed.

In [ ]:
# Your code here
